src/Identity/IdentityServer/UserDecryptionOptionsBuilder.cs

```diff
@@ -120,26 +124,32 @@ private async Task BuildTrustedDeviceOptions()
+            // except for their current device.
+            hasLoginApprovingDevice = allDevices.Any(d =>
+                d.Identifier != _device.Identifier &&
+                _loginApprovingClientTypes.TypesThatCanApprove.Contains(DeviceTypes.ToClientType(d.Type)));
         }
 
-        // Determine if user has manage reset password permission as post sso logic requires it for forcing users with this permission to set a MP
-        var hasManageResetPasswordPermission = false;
-        // when a user is being created via JIT provisioning, they will not have any orgs so we can't assume we will have orgs here
-        if (_currentContext.Organizations != null && _currentContext.Organizations.Any(o => o.Id == _ssoConfig.OrganizationId))
-        {
-            // TDE requires single org so grabbing first org & id is fine.
-            hasManageResetPasswordPermission = await _currentContext.ManageResetPassword(_ssoConfig!.OrganizationId);
-        }
-
-        // If sso configuration data is not null then I know for sure that ssoConfiguration isn't null
-        var organizationUser = await _organizationUserRepository.GetByOrganizationAsync(_ssoConfig.OrganizationId, _user.Id);
-
-        hasManageResetPasswordPermission |= organizationUser != null && (organizationUser.Type == OrganizationUserType.Owner || organizationUser.Type == OrganizationUserType.Admin);
-        // They are only able to be approved by an admin if they have enrolled is reset password
+        var organizationUser =
+            await _organizationUserRepository.GetByOrganizationAsync(_ssoConfig.OrganizationId, _user.Id);
+
+        // SSO users, Organizations, and Providers with Manage Account Recovery (reset password) permission,
+        // as well as all Owners and Admins, must set a Master Password.
+        var hasManageResetPasswordPermission =
+            // We need to interrogate the organizationUser from the repository to solve for an invited status.
+            // The claims won't exist at that time, so currentContext as it is written today won't accommodate for that.
+            // This is the edge case.
+            // TODO: PM-25668 proposes some refactor of the Organization acceptance processes to unify requirement for
+            // Organization acceptance; once it is implemented, we should be able to remove this Organization User check.
+            organizationUser?.GetPermissions() is { ManageResetPassword: true } ||
+            organizationUser?.Type is OrganizationUserType.Admin or OrganizationUserType.Owner ||
+            // For Organization User in Accepted/Confirmed status, claims will be available. CurrentContext will also
+            // offer Provider User information for that case. As MSP, Provider Users will require certain permissions like
+            // ManageResetPassword to perform their role effectively.
+            await _currentContext.ManageResetPassword(_ssoConfig!.OrganizationId);
```

Comment:
This comment also doesn't match the code. `CurrentContext` only contains users in the Accepted status.

Given that this is only here to check for the provider status, you can use a more precise `CurrentContext` method to check this:

```c#
await _currentContext.ProviderUserForOrgAsync(_ssoConfig!.OrganizationId);
```

This more clearly separates member and provider checks.